### PR TITLE
Klarna#177478237

### PR DESCRIPTION
**Pull request by patmbolger**
#### PT URL: https://www.pivotaltracker.com/story/show/177478237

---
## Ready for review:
@

### CODE REVIEW

File: app/models/company.rb [hunk 1]
```diff
@@ -127,7 +127,7 @@ def self.address_visible
 
   # TODO: yuck
   def self.with_members
-    where(id: CompanyApplication.where(shf_application: [ShfApplication.where(user: User.current_member)]).pluck(:company_id))
+    where(id: CompanyApplication.where(shf_application: [ShfApplication.where(user: User.members)]).pluck(:company_id))
```
Comment (patmbolger) [inline app/models/company.rb]: @weedySeaDragon I changed this `User` scope because I had failing tests (`problem-scenarios-pagination-view_all_companies`) that were failing because the users created for the test did not actually have membership status (when they should have).  We'll have to adjust certain test steps.

Comment (weedySeaDragon) [reply app/models/company.rb]: I'm confused by your statement:
"the users created for the test did not actually have membership status (when they should have)"

Does that mean that:
1.  you did not create current members?
or 
2.  you _did_ create current members but the system did not recognized them as _current members_ ?

If #1 then the solution is not to revert code so that it is incorrect (the query you changed will now no longer be correct in terms of either business rules or implementation), but to fix your tests.

If #2 then it's a bigger problem and you should create tests (feature or RSpec) that show the problem so that we're all clear on it and can fix it.

File: Gemfile [hunk 2]
```diff
@@ -5,6 +5,8 @@ gem 'dotenv-rails'
 gem 'rails', '~> 5.2'
 gem 'bootsnap', require: false
 
+gem 'mimemagic', '= 0.3.5' # https://dev.to/cseeman/what-s-up-with-mimemagic-breaking-everything-he1
```
Comment (weedySeaDragon) [inline Gemfile]: You should upgrade  your gem (and the Gemfile.lock) to  mimemagic (0.3.10).

File: config/initializers/klarna_service.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+if Rails.env.production?
+  KLARNA_API_AUTH_USERNAME  = ENV['SHF_KLARNA_API_AUTH_USERNAME']
+  KLARNA_API_AUTH_PASSWORD  = ENV['SHF_KLARNA_API_AUTH_PASSWORD']
+else
+  KLARNA_API_AUTH_USERNAME  = ENV['SHF_KLARNA_API_AUTH_USERNAME_DEV']
+  KLARNA_API_AUTH_PASSWORD  = ENV['SHF_KLARNA_API_AUTH_PASSWORD_DEV']
+end
+
+SHF_MEMBER_FEE   = 30000  # unit == one hundredth krona
```
Comment (weedySeaDragon) [inline config/initializers/klarna_service.rb:12]: We should really pull these 2 fees out into ENV files.
They shouldn't be buried in gem initialization files (either HIPS or KLARNA).
Can do this now or in a later PR.

(Just skimming thru your changes.)

Comment (patmbolger) [reply config/initializers/klarna_service.rb:12]: Not sure I agree.   These values are 1) relatively invariant (unless SHF changes fees every few years, but certainly not very often), and 2) there is no reason to use different  values in different environments.  The overhead of maintaining ENV vars, in different environments, is greater than the convenience of having them dynamically settable.

Comment (weedySeaDragon) [reply config/initializers/klarna_service.rb:12]: Klarna does not have anything to do with setting these values.  Klarna is not responsible for these values in any way.  Thus they should not be associated with Klarna or the Klarna API (or HIPS for that matter.)

If they are "relatively invariant" then how can it also be important to be "dynamically settable"?  

The fees are really the responsibility of the admin, just like other values that we have in the Admin Config.  (We also have some values in the ENV files that should really be in the Admin config.)

File: features/step_definitions/user_steps.rb [hunk 4]
```diff
@@ -44,6 +44,8 @@ def user_agrees_to_membership_guidelines(user)
     AdminOnly::UserChecklistFactory.create_member_guidelines_checklist_for(new_user)
     user_agrees_to_membership_guidelines(new_user) if user_agreed_to_membership_guidelines
     new_user
+
+    new_user.start_membership!(send_email: false) if new_user.member?
```
Comment (weedySeaDragon) [inline features/step_definitions/user_steps.rb]: should really query the membership state: `.current_member?` 
AASM gem creates all of the query/boolean methods based on the different states.
That ensures that all of the requirements are checked and everything is up to date.
(Same as with the ShfApplication, of course)

File: Gemfile (hunk 2 above)
Comment (patmbolger) [reply Gemfile]: @weedySeaDragon In order to use `0.3.10` I need to install a local mime DB on my laptop.  The recommended solution is to install `shared-mime-info` via brew (on Mac).  Do we have that package installed on DI? (I assume yes)

File: features/step_definitions/user_steps.rb (hunk 4 above)
Comment (patmbolger) [reply features/step_definitions/user_steps.rb]: Yes, I know, but that method fails when called in the context of `problem-scenarios-pagination-view_all_companies`.  For purposes of that test, I just need the users to appear to be "members" (and that test (line 179) doesn't care that the members actually have not paid their membership fee, and thus cannot be members in the real world).

The set up data is: 

```
And the following users exist:
      | email         | admin | member | agreed_to_membership_guidelines |
      | u1@mutts.com  |       | true   | true                            |
      | u2@mutts.com  |       | true   | true                            |
...
```
where we're using the "member" attribute to indicate that the user should _look_ like a member for test purposes.

Short of restructuring many of our test setups around the actual requirements for an object to meet criteria for a state, I think this is a reasonable compromise, as long as mocking (in effect) the actual requirements doe not affect the integrity of the specific test.

Comment (weedySeaDragon) [reply features/step_definitions/user_steps.rb]: I've already done all that work to change the tests.  That was a big part of what I had to do to implement the new membership states, the Membership class, and the business logic.

 When someone new joins the project and reads through the tests, yours could (probably would) mislead them in understanding both the requirements and the implementation.

You need to:
1. set the `membership_status.  That is now how we track the membership status (since there is more than just the binary 'yes a member / no, not a member' state.  
2. create a 'Membership' object

You can see an example in the `features/user_account/member-home-account-page.feature`
There's a column for `membership_status` and then a section in the Background for a membership: `And the following memberships exist`  (I made sure to set this up so you can copy the first few columns from the payments set-up.)

File: Gemfile (hunk 2 above)
Comment (weedySeaDragon) [reply Gemfile]: I think we should use upgrade Rails to 5.2.6 -- the version they just released that does not use mime-magic.  (They revised what was needed into a gem they named 'marcel' (!) ).
We'll need to update soon just to take care of the security issues anyway.  

I tried the upgrade and ran the tests and _of course_ there was a problem.  Upgrading to that version Rails also upgrades the i18n gem to v.1.8.10.  The i18n gem changed the translate method: they no longer accept an empty Hash as an argument (in preparation for ruby 3.0, in part).  The only place we pass that in is with the Cucumber features:  it is in a method in `features/support/env.rb`:  `i18n_content`  .  (Like all problems, it took hours to find and minutes to fix.)
I have a fix for it but need to put it into a PR, etc.  Will do that tomorrow.

In doing that fix, I may have found a problem with the `features/view-site-in-lang-i18n-l10n.feature`   I need to look at that again tomorrow.

I'll put the PR up tomorrow (Tuesday).  If you want to get it working in the meantime, here's the method that will change:

```
# As of version 1.8.6, I18n.translate will no longer accept a Hash as the 2nd argument.
# See I18n.translate
def i18n_content(content, translation_params = {})
  I18n.t(content, **({locale: I18n.locale}.merge(translation_params)) )
end
```

Comment (patmbolger) [reply Gemfile]: @weedySeaDragon Yes, I had changed that method `i18n_content` in this PR because it broke for me - the 2nd argument to that method had been wrong for some time.

I was aware of the Rails upgrade that avoided the mimemagic issue, but did not want to add that to this story.  However, if you think it's a viable approach I'm all for it.

File: app/controllers/payments_controller.rb [hunk 5]
```diff
@@ -5,67 +5,134 @@ class PaymentsController < ApplicationController
...
+
     # The user wants to pay a fee (e.g. membership fee or branding fee)
     payment_type = params[:type]
     user_id      = params[:user_id]
-    company_id   = params[:company_id]
+    company_id   = params[:company_id] # if h-branding fee to be paid
 
     authorize Payment.new(user_id: user_id)
 
     if payment_type == Payment::PAYMENT_TYPE_MEMBER
```
Comment (weedySeaDragon) [inline app/controllers/payments_controller.rb]: Don't use the constant from another class.  Use the method.

File: app/controllers/payments_controller.rb [hunk 6]
```diff
@@ -5,67 +5,134 @@ class PaymentsController < ApplicationController
...
-      entity_id = user_id
       start_date, expire_date = User.next_membership_payment_dates(user_id)
       paid_item = I18n.t('payment.payment_type.member_fee')
+      item_price = SHF_MEMBER_FEE
```
Comment (weedySeaDragon) [inline app/controllers/payments_controller.rb:25]: This should be an AppConfiguration value, a should the branding fee amount.  (We can do it in a separate PR)

File: app/controllers/payments_controller.rb [hunk 7]
```diff
@@ -5,67 +5,134 @@ class PaymentsController < ApplicationController
...
+                               company_id: company_id,
+                               status: Payment.order_to_payment_status(nil),
+                               payment_processor: Payment::PAYMENT_PROCESSOR_KLARNA,
+                               start_date: start_date,
+                               expire_date: expire_date)
+
+    payment_data = { id: @payment.id,
+                     user_id: user_id,
+                     type: payment_type,
+                     currency: 'SEK',
```
Comment (weedySeaDragon) [inline app/controllers/payments_controller.rb]: make the currency a constant

File: app/controllers/payments_controller.rb [hunk 8]
```diff
@@ -5,67 +5,134 @@ class PaymentsController < ApplicationController
...
 
-  rescue RuntimeError, HTTParty::Error, ActiveRecord::RecordInvalid  => exc
-    @payment.destroy if @payment.persisted?
+    log_klarna_activity('Create Order', 'info', @payment.id, klarna_id)
 
-    log_hips_activity('create order', 'error', nil, @hips_id, exc)
+    @html_snippet = klarna_order['html_snippet']
 
-    log_hips_activity('create order', 'error', nil, @hips_id, exc.cause)
+  rescue RuntimeError, HTTParty::Error, ActiveRecord::RecordInvalid  => exc
```
Comment (weedySeaDragon) [inline app/controllers/payments_controller.rb]: RuntimeErrors shouldn't be explicitly caught here. 
Use separate rescue clauses for the HTTParty & invalid record errors so that each can have specific, informative errors with specific text.

File: app/controllers/payments_controller.rb [hunk 9]
```diff
@@ -5,67 +5,134 @@ class PaymentsController < ApplicationController
...
+
+    redirect_back fallback_location: root_path
+  end
+
+  def confirmation
+    # https://developers.klarna.com/documentation/klarna-checkout/in-depth/confirm-purchase
+
+    payment_id = params[:id]
+    klarna_id = params[:klarna_id]
+    raise 'No Klarna order ID' unless klarna_id
```
Comment (weedySeaDragon) [inline app/controllers/payments_controller.rb:80]: Make this a simple error class, ex:
```
class KlarnaError < StandardError; end
class NoKlarnaOrderid < KlarnaError; end
etc
```
It's considered a best practice to create specific Error classes so that you can catch the most specific error possible, so that you can associate specific helpful messages (e.g. for logs) with each, and so that development/testing/maintenance is better encapsulated and SOLID.

File: app/controllers/payments_controller.rb [hunk 10]
```diff
@@ -5,67 +5,134 @@ class PaymentsController < ApplicationController
...
+                                        user_path(params[:user_id]))
+
+    helpers.flash_message(:notice, t('.success_html',
+                                     account_page_link: account_page_link))
+
+    log_klarna_activity('Order Confirmation', 'info', payment_id, klarna_id)
+
+    @html_snippet = klarna_order['html_snippet']
+
+  rescue RuntimeError, HTTParty::Error, ActiveRecord::RecordInvalid => exc
```
Comment (weedySeaDragon) [inline app/controllers/payments_controller.rb]: same comment about errors

File: app/controllers/payments_controller.rb [hunk 11]
```diff
@@ -5,67 +5,134 @@ class PaymentsController < ApplicationController
...
+    notify_slack_of_exception(exc, __method__)
+
+    helpers.flash_message(:alert, t('payments.create.something_wrong',
+                                    admin_email: ENV['SHF_REPLY_TO_EMAIL']))
+
     redirect_back fallback_location: root_path
   end
 
+  def klarna_push
+    # This is the klarna "push" action.  It is the "fallback" action in case
```
Comment (weedySeaDragon) [inline app/controllers/payments_controller.rb]: Put comments above method signatures.  Yard and other documenters expect it.  (perhaps even ruby style)

File: app/controllers/payments_controller.rb [hunk 12]
```diff
@@ -5,67 +5,134 @@ class PaymentsController < ApplicationController
...
+    # Fetch the order (Order Management API) and check if "captured_amount" is zero:
+    #   If so, do nothing (order has been acknowledged and order amount captured).
+    #   Otherwise, perform same actions as for "confirmation" action.
+
+    klarna_id = params[:klarna_id]
+    payment_id = params[:id]
+
+    klarna_order = KlarnaService.get_order(klarna_id)
+
+    return if klarna_order['captured_amount'] != 0
```
Comment (weedySeaDragon) [inline app/controllers/payments_controller.rb:119]: what does this do?  (doc in method info, perhaps?)

File: app/controllers/payments_controller.rb [hunk 13]
```diff
@@ -5,67 +5,134 @@ class PaymentsController < ApplicationController
...
+
+    handle_order_confirmation(klarna_id, payment_id)
+
+    log_klarna_activity('Webhook', 'info', payment_id, klarna_id)
+
+  rescue RuntimeError, HTTParty::Error, ActiveRecord::RecordInvalid => exc
```
Comment (weedySeaDragon) [inline app/controllers/payments_controller.rb]: same comment about errors

File: app/controllers/payments_controller.rb [hunk 14]
```diff
@@ -5,67 +5,134 @@ class PaymentsController < ApplicationController
...
+
+    notify_slack_of_exception(exc, __method__)
+
+    helpers.flash_message(:alert, t('payments.create.something_wrong',
+                                    admin_email: ENV['SHF_REPLY_TO_EMAIL']))
+
     redirect_back fallback_location: root_path
   end
 
+  def klarna_push
```
Comment (weedySeaDragon) [inline app/controllers/payments_controller.rb]: This duplicates a lot with `confirmation`.  DRY up

File: app/controllers/payments_controller.rb [hunk 15]
```diff
@@ -95,52 +162,61 @@ def webhook
...
-  def redirect_on_payment_success_or_error(payment)
+    KlarnaService.acknowledge_order(klarna_id)
 
-    if payment.payment_type == Payment::PAYMENT_TYPE_MEMBER
-      redirect_to user_path(params[:user_id])
-    else
-      redirect_to company_path(payment.company_id)
-    end
+    payment = Payment.find(payment_id)
+    payment.update_attribute(:status,
```
Comment (weedySeaDragon) [inline app/controllers/payments_controller.rb]: why `update_attribute` instead of just `update` (which will also validate)?

File: app/controllers/payments_controller.rb [hunk 16]
```diff
@@ -5,67 +5,134 @@ class PaymentsController < ApplicationController
...
+    @payment.klarna_id = klarna_id
+    @payment.status = Payment.order_to_payment_status(klarna_order['status'])
     @payment.save!
 
-  rescue RuntimeError, HTTParty::Error, ActiveRecord::RecordInvalid  => exc
-    @payment.destroy if @payment.persisted?
+    log_klarna_activity('Create Order', 'info', @payment.id, klarna_id)
 
-    log_hips_activity('create order', 'error', nil, @hips_id, exc)
+    @html_snippet = klarna_order['html_snippet']
```
Comment (weedySeaDragon) [inline app/controllers/payments_controller.rb:61]: `exc` is not a meaningful/readable variable name

File: app/controllers/payments_controller.rb [hunk 17]
```diff
@@ -95,52 +162,61 @@ def webhook
...
+    # payment settlement can occur)
+    KlarnaService.capture_order(klarna_id, klarna_order['order_amount'])
+
+    klarna_order
   end
 
-  def log_hips_activity(activity, severity, payment_id, hips_id, exc=nil)
-    ActivityLogger.open(HIPS_LOG, 'HIPS_API', activity, false) do |log|
+  def log_klarna_activity(activity, severity, payment_id, klarna_id, exc=nil)
+    ActivityLogger.open(KLARNA_LOG, 'Payments', activity, false) do |log|
```
Comment (weedySeaDragon) [inline app/controllers/payments_controller.rb]: Use `LogfileNamer` to create the standardized name for the log file. (Also means that the constant is not needed in the initializer.)

File: app/controllers/payments_controller.rb [hunk 18]
```diff
@@ -95,52 +162,61 @@ def webhook
...
+    # Capture the order in Klarna (indicates the order has been filled and
+    # payment settlement can occur)
+    KlarnaService.capture_order(klarna_id, klarna_order['order_amount'])
+
+    klarna_order
   end
 
-  def log_hips_activity(activity, severity, payment_id, hips_id, exc=nil)
-    ActivityLogger.open(HIPS_LOG, 'HIPS_API', activity, false) do |log|
+  def log_klarna_activity(activity, severity, payment_id, klarna_id, exc=nil)
```
Comment (weedySeaDragon) [inline app/controllers/payments_controller.rb]: `exc` not a meaningful/readable name

File: app/controllers/payments_controller.rb [hunk 19]
```diff
@@ -95,52 +162,61 @@ def webhook
...
+                                   type: payment_type)
 
-    urls[:error]   = payment_error_url(user_id: user_id,
-                                       company_id: company_id,
-                                       id: payment_id,
-                                       payment_type: payment_type)
+    urls[:confirmation] = payment_confirmation_url(id: payment_id,
+                                                   user_id: user_id,
+                                                   disable_language_change: true,
+                                                   klarna_id: '{checkout.order.id}')
```
Comment (weedySeaDragon) [inline app/controllers/payments_controller.rb:208]: `klarna_id` is set to the string `{checkout.order.id}` ?
or should this be interpolated? `"#{checkout.order.id}"`

File: app/controllers/payments_controller.rb [hunk 20]
```diff
@@ -95,52 +162,61 @@ def webhook
...
 
-    urls[:webhook] = (SHF_WEBHOOK_HOST || root_url) + payment_webhook_path.sub('/en', '')
+    urls[:push] = (SHF_WEBHOOK_HOST || root_url) +
+                  payment_klarna_push_path(id: payment_id,
+                                           klarna_id: '{checkout.order.id}').sub('/en', '')
```
Comment (weedySeaDragon) [inline app/controllers/payments_controller.rb:212]: since you're reusing the value (checkout.order.id), assign it to a variable (DRY)

File: app/services/klarna_service.rb [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+# https://developers.klarna.com/api/#checkout-api
+class KlarnaService
+  require 'httparty'
+
+  SUCCESS_CODES = [200, 201, 202, 204].freeze
+
+  def self.create_order(payment_data)
+
+    raise 'Invalid payment type' unless
+      payment_data[:type] == Payment::PAYMENT_TYPE_MEMBER ||
```
Comment (weedySeaDragon) [inline app/services/klarna_service.rb]: don't access constants in another class directly.  Use methods provided

File: app/services/klarna_service.rb [hunk 22]
```diff
@@ -0,0 +1,165 @@
+# Interface to Klarna Checkout service API
+# https://developers.klarna.com/api/#checkout-api
+class KlarnaService
```
Comment (weedySeaDragon) [inline app/services/klarna_service.rb]: Lots of duplicated code in methods; DRY.
Use simple error classes, ex:
```
class KlarnaServiceError < StandardError; end
class KlarnaInvalidPaymentTypeError < KlarnaServiceError; end
class Klarna401UnauthorizedError < KlarnaServiceError; end
class Klarna404NotFoundError < KlarnaServiceError; end
class KlarnaResponseError < KlarnaServiceError; end
class KlarnaUnknownResponseError < KlarnaServiceError; end
....
```
Then rescue those errors specifically.  Don't rescue `RuntimeError`

use AppConfig variable for terms URL 'https://sverigeshundforetagare.se/aterbetalningsvillkor/',

File: app/services/klarna_service.rb [hunk 23]
```diff
@@ -0,0 +1,165 @@
...
+  require 'httparty'
+
+  SUCCESS_CODES = [200, 201, 202, 204].freeze
+
```
Comment (weedySeaDragon) [inline app/services/klarna_service.rb]: Use constants for values in the `options`.  Ex:
```
ALLOWED_CUSTOMER_TYPES = %w[person organization].freeze
  BOOTSTRAP_PRIMARY_COLOR = '#003a78'.freeze
  WHITE = '#ffffff'.freeze
  DARK_BROWN_LINK_COLOR = '#232525'.freeze

  PAYMENT_ORDER_OPTIONS = {
    color_button: BOOTSTRAP_PRIMARY_COLOR,
    color_button_text: WHITE,
    color_header: DARK_BROWN_LINK_COLOR,
    allowed_customer_types: ALLOWED_CUSTOMER_TYPES
  }
  KLARNA_JSON_TYPE = 'application/vnd.klarna.internal.emd-v2+json'.freeze
  ORDER_TYPE = 'digital'.freeze
```

File: app/services/klarna_service.rb [hunk 24]
```diff
@@ -0,0 +1,165 @@
...
+  def self.create_order(payment_data)
+
+    raise 'Invalid payment type' unless
+      payment_data[:type] == Payment::PAYMENT_TYPE_MEMBER ||
+      payment_data[:type] == Payment::PAYMENT_TYPE_BRANDING
+
+    response = HTTParty.post(KLARNA_CHECKOUT_URL,
```
Comment (weedySeaDragon) [inline app/services/klarna_service.rb]: If KLARNA_CHECKOUT_URL is only used in this file, define it here.  (readability/maintenance, scope)

File: app/services/klarna_service.rb [hunk 25]
```diff
@@ -0,0 +1,165 @@
...
+                             basic_auth: auth,
+                             headers: { 'Content-Type' => 'application/json' },
+                             body: order_json(payment_data))
+
+    return response.parsed_response if SUCCESS_CODES.include?(response.code)
+
+    # Wrap cause exception within HTTP error exception so both appear in log
+    begin
+      process_api_error(response)
+    rescue RuntimeError
```
Comment (weedySeaDragon) [inline app/services/klarna_service.rb]: Should try to capture specific errors and then default to a generic StandardError if no others were matched.

**Review by weedySeaDragon — CHANGES_REQUESTED** [review]
Logic looks good.
DRY, constants, and error handling need to be addressed.  (See my specific comments in the code.)
(+ my previous comments in the PR)
Gemfile.lock:  Looks like you have updated _all_ gems.  There are some that have major changes, some with minor changes (according to the version numbers), and a few that even went to earlier versions.  Was this intentional?  I'd recommend doing this in a separate PR.

File: app/controllers/payments_controller.rb (hunk 8 above)
Comment (patmbolger) [reply app/controllers/payments_controller.rb]: I don't have a need to setting exception-class-specific text here.  The exception(s) will be unwound (each exception and cause in the error stack) in `log_klarna_activity` and `notify_slack_of_exception`, with required detail.

Also, I am raising Runtime errors in `KlarnaService`.

File: app/controllers/payments_controller.rb (hunk 14 above)
Comment (patmbolger) [reply app/controllers/payments_controller.rb]: I don't see a lot of duplication (?)

File: app/controllers/payments_controller.rb (hunk 11 above)
Comment (patmbolger) [reply app/controllers/payments_controller.rb]: OK

File: app/controllers/payments_controller.rb (hunk 12 above)
Comment (patmbolger) [reply app/controllers/payments_controller.rb:119]: This return with no action is explained in comment above the method definition.

File: app/controllers/payments_controller.rb (hunk 15 above)
Comment (patmbolger) [reply app/controllers/payments_controller.rb]: I'm only updating the one attribute.  `update` would require all attributes to be present in the hash argument.

I don't actually validate the `status`, other than its presence is required. I suppose I could log a message to log file and slack if we encounter an unknown status.  I'll add that.

File: app/controllers/payments_controller.rb (hunk 8 above)
Comment (weedySeaDragon) [reply app/controllers/payments_controller.rb]: I'm not talking necessarily about the text.  I'm talking about using Error classes that are specific to Klarna interaction.  The text is helpful if you're looking at logs.  Using classes is what makes understanding, maintaining and extending the code easier in the future. It makes it easy for any code written in the future to detect and rescue (capture) them.  Future code simply needs to rescue`KlarnaError` to separate out those errors from a problem arising from something else.  
IOW, a future developer won't have do put in more conditional code to see if an error is actually from something with Klarna vs. something else.
It also makes testing specific and easier.  Your RSpec tests can create the specific conditions that should (or not) raise specific errors.  

Avdi Grimm has an excellent small book on Exception handling that is considered best practices.

File: app/controllers/payments_controller.rb (hunk 12 above)
Comment (weedySeaDragon) [reply app/controllers/payments_controller.rb:119]: doh! I totally missed that.  thanks.

File: app/controllers/payments_controller.rb (hunk 15 above)
Comment (weedySeaDragon) [reply app/controllers/payments_controller.rb]: `update` doesn't update all of the attributes -- only the ones you pass in. https://guides.rubyonrails.org/v5.2/active_record_basics.html#update
It's the standard method for persisting attribute values.

File: app/controllers/payments_controller.rb (hunk 14 above)
Comment (weedySeaDragon) [reply app/controllers/payments_controller.rb]: Your rescue clauses are duplicates.  At the least you should have a method that returns the list of errors you're checking/rescuing and then use that instead of always listing the same classes.
You are also always calling `log_klarna_activity` and `notify_slack_of_exception`.  Create 1 method that calls them both and then just use that 1 method.

In general the methods have very similar designs (overall pattern). That is an indicator that there may be common behavior that can be refactored into common methods that can be called and/or that a more general 'outer' method can be used with a `yield` statement to handle  behavior specific to each common original method.

File: app/controllers/payments_controller.rb (hunk 19 above)
Comment (patmbolger) [inline app/controllers/payments_controller.rb:208]: No ... Klarna will replace that string with actual ID when that webhook URL is invoked.

File: app/controllers/payments_controller.rb [hunk 26]
```diff
@@ -95,52 +162,61 @@ def webhook
...
   end
+
+  def notify_slack_of_exception(exc, method_sym)
+    class_and_method = "#{self.class.name}\##{method_sym}"
+
+    [exc, exc.cause].each do |this_exc|
+      break unless this_exc
+      SHFNotifySlack.failure_notification(class_and_method, text: this_exc.message)
+    end
+  end
```
Comment (patmbolger) [inline app/controllers/payments_controller.rb:223]: I wonder if we want to add this method to `SHFNotifySlack`.

Comment (weedySeaDragon) [reply app/controllers/payments_controller.rb:223]: This seems outside the scope/responsibility of `SHFNotifySlack`.  
 Its responsibility is to notify Slack, not to format or gather the information. 
`SHFNotifySlack` expects all of the information (exception, whatever) to be sent to it.  That has to take place before it gets sent to `SHFNotifySlack`.

I can see having a class that takes and exception, gets the info from it (perhaps as you've done) and logs and sends it to `SHFNotifySlack`.

Also:  (1) Why are you handling the exception and its cause like this?
(2) you should have an RSpec test -- that will explain it (as well as test it, obv.)
You are always doing that: logging and notifying slack; you can DRY up code by having 1 method handle those two calls.

File: app/controllers/payments_controller.rb (hunk 19 above)
Comment (weedySeaDragon) [reply app/controllers/payments_controller.rb:208]: Can you put that in a comment, please?  I won't be the only one to wonder that when looking at the code. :-)

File: app/controllers/payments_controller.rb (hunk 26 above)
Comment (patmbolger) [reply app/controllers/payments_controller.rb:223]: This is a convenient way to unpack the exception stack and post the whole "story" to slack.
The two methods have different signatures and a combined method would thus not make sense.

Comment (weedySeaDragon) [reply app/controllers/payments_controller.rb:223]: We should have the same information (the cause/ the whole story) recorded to the log so that we never lose that information.

Does seem like a method that takes an exception and returns a string with the "whole story" would be a really good method to have and be available to the whole system.  (Then that 1 method could be called so it would be written out to where-ever is needed, i.e. the log, Slack,...)
Don't know if this would just be in ApplicationController, or where.
(Separate PR to be sure!)
